## 4.8 Task - Aggregating Variables

## Find the mean orders for each department in the merged dataframe.
## Create loyalty flag for existing customers.
## Find the mean of product prices by loyalty group.
## Create spending flag for each user based on average price of products across orders.
## Create order frequency flag that indicates users' ordering behavior.

In [2]:
## Import libraries

import pandas as pd
import numpy as np
import os

In [3]:
#importing the merged data
path = r'/Users/frankamoboateng/Downloads/18.07.23 instacart Basket Analysis'

In [4]:
path

'/Users/frankamoboateng/Downloads/18.07.23 instacart Basket Analysis'

In [5]:
# importing ords_prods_merged
df_ords_prods_merged = pd.read_pickle(os.path.join(path,'02. DATA','PREPARED DATA','orders_products_merged_updated.pkl')) 
                    

In [6]:
#importing department wrangled file
df_dept = pd.read_csv(os.path.join(path,'02. DATA','PREPARED DATA','department.wrangled.csv'), 
                               index_col = False)

In [7]:
df_dept

Unnamed: 0       department
0            1           frozen
1            2            other
2            3           bakery
3            4          produce
4            5          alcohol
5            6    international
6            7        beverages
7            8             pets
8            9  dry goods pasta
9           10             bulk
10          11    personal care
11          12     meat seafood
12          13           pantry
13          14        breakfast
14          15     canned goods
15          16       dairy eggs
16          17        household
17          18           babies
18          19           snacks
19          20             deli
20          21          missing

In [8]:
#inspecting the file 
df_ords_prods_merged.shape

(32434212, 21)

In [9]:
#checking few data
df_ords_prods_merged.head()

Unnamed: 0  Unnamed: 0_x  product_id                product_name  aisle_id  \
0           0             0           1  Chocolate Sandwich Cookies        61   
1           1             0           1  Chocolate Sandwich Cookies        61   
2           2             0           1  Chocolate Sandwich Cookies        61   
3           3             0           1  Chocolate Sandwich Cookies        61   
4           4             0           1  Chocolate Sandwich Cookies        61   

   department_id  prices  Unnamed: 0.2  Unnamed: 0.1  Unnamed: 0_y  ...  \
0             19     5.8         16802          1987          1987  ...   
1             19     5.8         16806          1989          1989  ...   
2             19     5.8        106710         11433         11433  ...   
3             19     5.8        113536         12198         12198  ...   
4             19     5.8        113575         12200         12200  ...   

   user_id  order_number  orders_day_of_week  order_hour_of_day  \
0      138            28                   6                 11   
1      138            30                   6                 17   
2      709             2                   0                 21   
3      764             1                   3                 13   
4      764             3                   4                 17   

   days_since_prior_order  add_to_cart_order  reordered  _merge  \
0                     3.0                  5          0    both   
1                    20.0                  1          1    both   
2                     6.0                 20          0    both   
3                     1.0                 10          0    both   
4                     9.0                 11          1    both   

      busiest_day busiest_period_of_day  
0  Regularly busy           Most orders  
1  Regularly busy        Average orders  
2    Busiest days        Average orders  
3    Slowest days           Most orders  
4    Slowest days        Average orders  

[5 rows x 21 columns]

In [10]:
df_ords_prods_merged.columns

Index(['Unnamed: 0', 'Unnamed: 0_x', 'product_id', 'product_name', 'aisle_id',
       'department_id', 'prices', 'Unnamed: 0.2', 'Unnamed: 0.1',
       'Unnamed: 0_y', 'order_id', 'user_id', 'order_number',
       'orders_day_of_week', 'order_hour_of_day', 'days_since_prior_order',
       'add_to_cart_order', 'reordered', '_merge', 'busiest_day',
       'busiest_period_of_day'],
      dtype='object')

# finding the mean orders for each deparment

In [11]:
# Group by "product_name" 
df_ords_prods_merged.groupby('product_name')

In [12]:
# Spliting the data into groups by "department_id" and then calculate the mean for each "order_number" column
df_ords_prods_merged.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457687
2                17.277920
3                17.179756
4                17.811403
5                15.213779
6                16.439806
7                17.225773
8                15.340520
9                15.895474
10               20.197148
11               16.170828
12               15.887622
13               16.583304
14               16.757377
15               16.165037
16               17.663250
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## The mean orders show less variation when calculated using the entire dataframe compared to a smaller subset. In the subset, the minimum mean order is 11.3, whereas in the dataframe, it is 15.2. The maximum mean order in the subset is 19.5, while in the dataframe, it is 20.1 (excluding department 21, the "missing" category).

# # Creating loyalty flag for the existing customers

In [13]:
# Group data by "user_id" column, generate maximum orders for each user, 
#create "max_order column" for aggregation results
df_ords_prods_merged['max_order'] = df_ords_prods_merged.groupby(['user_id'])['order_number'].transform(np.max)

In [14]:
# Check the data grouping for accuracy
pd.options.display.max_rows = None

In [15]:
df_ords_prods_merged.head(100)

Unnamed: 0  Unnamed: 0_x  product_id                product_name  \
0            0             0           1  Chocolate Sandwich Cookies   
1            1             0           1  Chocolate Sandwich Cookies   
2            2             0           1  Chocolate Sandwich Cookies   
3            3             0           1  Chocolate Sandwich Cookies   
4            4             0           1  Chocolate Sandwich Cookies   
5            5             0           1  Chocolate Sandwich Cookies   
6            6             0           1  Chocolate Sandwich Cookies   
7            7             0           1  Chocolate Sandwich Cookies   
8            8             0           1  Chocolate Sandwich Cookies   
9            9             0           1  Chocolate Sandwich Cookies   
10          10             0           1  Chocolate Sandwich Cookies   
11          11             0           1  Chocolate Sandwich Cookies   
12          12             0           1  Chocolate Sandwich Cookies   
13          13             0           1  Chocolate Sandwich Cookies   
14          14             0           1  Chocolate Sandwich Cookies   
15          15             0           1  Chocolate Sandwich Cookies   
16          16             0           1  Chocolate Sandwich Cookies   
17          17             0           1  Chocolate Sandwich Cookies   
18          18             0           1  Chocolate Sandwich Cookies   
19          19             0           1  Chocolate Sandwich Cookies   
20          20             0           1  Chocolate Sandwich Cookies   
21          21             0           1  Chocolate Sandwich Cookies   
22          22             0           1  Chocolate Sandwich Cookies   
23          23             0           1  Chocolate Sandwich Cookies   
24          24             0           1  Chocolate Sandwich Cookies   
25          25             0           1  Chocolate Sandwich Cookies   
26          26             0           1  Chocolate Sandwich Cookies   
27          27             0           1  Chocolate Sandwich Cookies   
28          28             0           1  Chocolate Sandwich Cookies   
29          29             0           1  Chocolate Sandwich Cookies   
30          30             0           1  Chocolate Sandwich Cookies   
31          31             0           1  Chocolate Sandwich Cookies   
32          32             0           1  Chocolate Sandwich Cookies   
33          33             0           1  Chocolate Sandwich Cookies   
34          34             0           1  Chocolate Sandwich Cookies   
35          35             0           1  Chocolate Sandwich Cookies   
36          36             0           1  Chocolate Sandwich Cookies   
37          37             0           1  Chocolate Sandwich Cookies   
38          38             0           1  Chocolate Sandwich Cookies   
39          39             0           1  Chocolate Sandwich Cookies   
40          40             0           1  Chocolate Sandwich Cookies   
41          41             0           1  Chocolate Sandwich Cookies   
42          42             0           1  Chocolate Sandwich Cookies   
43          43             0           1  Chocolate Sandwich Cookies   
44          44             0           1  Chocolate Sandwich Cookies   
45          45             0           1  Chocolate Sandwich Cookies   
46          46             0           1  Chocolate Sandwich Cookies   
47          47             0           1  Chocolate Sandwich Cookies   
48          48             0           1  Chocolate Sandwich Cookies   
49          49             0           1  Chocolate Sandwich Cookies   
50          50             0           1  Chocolate Sandwich Cookies   
51          51             0           1  Chocolate Sandwich Cookies   
52          52             0           1  Chocolate Sandwich Cookies   
53          53             0           1  Chocolate Sandwich Cookies   
54          54             0           1  Ch

In [16]:
# Creating flags to assign "loyalty" labels to user ID based on max order value
df_ords_prods_merged.loc[df_ords_prods_merged['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [17]:
df_ords_prods_merged.loc[(df_ords_prods_merged['max_order'] <= 40) & (df_ords_prods_merged['max_order'] > 10), 
                     'loyalty_flag'] = 'Regular customer'

In [18]:
df_ords_prods_merged.loc[df_ords_prods_merged['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [19]:
# Checking frequencies of loyalty flags for accuracy
df_ords_prods_merged['loyalty_flag'].value_counts(dropna = False)

Regular customer    15891077
Loyal customer      10293737
New customer         6249398
Name: loyalty_flag, dtype: int64

# Finding the mean of product prices by loyalty groups

In [20]:
# Group by "prices" 
df_ords_prods_merged.groupby('prices')

In [21]:
# Spliting the data into groups by "loyalty_flag" and then calculate the mean for each "prices" column
df_ords_prods_merged.groupby('loyalty_flag').agg({'prices': ['mean']})

prices
                       mean
loyalty_flag               
Loyal customer    10.388747
New customer      13.294370
Regular customer  12.496203

# Creating spending flag for each user based on average price of products across orders

In [22]:
# Group data by "user_id" column, generate mean product price for each user,
#create "mean_price" for aggregation results
df_ords_prods_merged['mean_price'] = df_ords_prods_merged.groupby(['user_id'])['prices'].transform(np.mean)

In [23]:
# Check grouping
df_ords_prods_merged.head(100)

Unnamed: 0  Unnamed: 0_x  product_id                product_name  \
0            0             0           1  Chocolate Sandwich Cookies   
1            1             0           1  Chocolate Sandwich Cookies   
2            2             0           1  Chocolate Sandwich Cookies   
3            3             0           1  Chocolate Sandwich Cookies   
4            4             0           1  Chocolate Sandwich Cookies   
5            5             0           1  Chocolate Sandwich Cookies   
6            6             0           1  Chocolate Sandwich Cookies   
7            7             0           1  Chocolate Sandwich Cookies   
8            8             0           1  Chocolate Sandwich Cookies   
9            9             0           1  Chocolate Sandwich Cookies   
10          10             0           1  Chocolate Sandwich Cookies   
11          11             0           1  Chocolate Sandwich Cookies   
12          12             0           1  Chocolate Sandwich Cookies   
13          13             0           1  Chocolate Sandwich Cookies   
14          14             0           1  Chocolate Sandwich Cookies   
15          15             0           1  Chocolate Sandwich Cookies   
16          16             0           1  Chocolate Sandwich Cookies   
17          17             0           1  Chocolate Sandwich Cookies   
18          18             0           1  Chocolate Sandwich Cookies   
19          19             0           1  Chocolate Sandwich Cookies   
20          20             0           1  Chocolate Sandwich Cookies   
21          21             0           1  Chocolate Sandwich Cookies   
22          22             0           1  Chocolate Sandwich Cookies   
23          23             0           1  Chocolate Sandwich Cookies   
24          24             0           1  Chocolate Sandwich Cookies   
25          25             0           1  Chocolate Sandwich Cookies   
26          26             0           1  Chocolate Sandwich Cookies   
27          27             0           1  Chocolate Sandwich Cookies   
28          28             0           1  Chocolate Sandwich Cookies   
29          29             0           1  Chocolate Sandwich Cookies   
30          30             0           1  Chocolate Sandwich Cookies   
31          31             0           1  Chocolate Sandwich Cookies   
32          32             0           1  Chocolate Sandwich Cookies   
33          33             0           1  Chocolate Sandwich Cookies   
34          34             0           1  Chocolate Sandwich Cookies   
35          35             0           1  Chocolate Sandwich Cookies   
36          36             0           1  Chocolate Sandwich Cookies   
37          37             0           1  Chocolate Sandwich Cookies   
38          38             0           1  Chocolate Sandwich Cookies   
39          39             0           1  Chocolate Sandwich Cookies   
40          40             0           1  Chocolate Sandwich Cookies   
41          41             0           1  Chocolate Sandwich Cookies   
42          42             0           1  Chocolate Sandwich Cookies   
43          43             0           1  Chocolate Sandwich Cookies   
44          44             0           1  Chocolate Sandwich Cookies   
45          45             0           1  Chocolate Sandwich Cookies   
46          46             0           1  Chocolate Sandwich Cookies   
47          47             0           1  Chocolate Sandwich Cookies   
48          48             0           1  Chocolate Sandwich Cookies   
49          49             0           1  Chocolate Sandwich Cookies   
50          50             0           1  Chocolate Sandwich Cookies   
51          51             0           1  Chocolate Sandwich Cookies   
52          52             0           1  Chocolate Sandwich Cookies   
53          53             0           1  Chocolate Sandwich Cookies   
54          54             0           1  Ch

In [24]:
# Creating flags to assign "spending" labels to user ID based on mean product price
df_ords_prods_merged.loc[df_ords_prods_merged['mean_price'] < 10, 'spending_flag'] = 'Low spender'

In [25]:
df_ords_prods_merged.loc[df_ords_prods_merged['mean_price'] >= 10, 'spending_flag'] = 'High spender'

In [26]:
# Check frequencies of spending flags for accuracy
df_ords_prods_merged['spending_flag'].value_counts(dropna = False)

Low spender     31798751
High spender      635461
Name: spending_flag, dtype: int64

# Creating order frequency flag that indicates users' ordering behavior

In [27]:
# Group data by "user_id" column, generate median product of "days since prior order, 
#create "median_days" for aggregation results
df_ords_prods_merged['median_days'] = df_ords_prods_merged.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [28]:
# Check grouping
df_ords_prods_merged.head(100)

Unnamed: 0  Unnamed: 0_x  product_id                product_name  \
0            0             0           1  Chocolate Sandwich Cookies   
1            1             0           1  Chocolate Sandwich Cookies   
2            2             0           1  Chocolate Sandwich Cookies   
3            3             0           1  Chocolate Sandwich Cookies   
4            4             0           1  Chocolate Sandwich Cookies   
5            5             0           1  Chocolate Sandwich Cookies   
6            6             0           1  Chocolate Sandwich Cookies   
7            7             0           1  Chocolate Sandwich Cookies   
8            8             0           1  Chocolate Sandwich Cookies   
9            9             0           1  Chocolate Sandwich Cookies   
10          10             0           1  Chocolate Sandwich Cookies   
11          11             0           1  Chocolate Sandwich Cookies   
12          12             0           1  Chocolate Sandwich Cookies   
13          13             0           1  Chocolate Sandwich Cookies   
14          14             0           1  Chocolate Sandwich Cookies   
15          15             0           1  Chocolate Sandwich Cookies   
16          16             0           1  Chocolate Sandwich Cookies   
17          17             0           1  Chocolate Sandwich Cookies   
18          18             0           1  Chocolate Sandwich Cookies   
19          19             0           1  Chocolate Sandwich Cookies   
20          20             0           1  Chocolate Sandwich Cookies   
21          21             0           1  Chocolate Sandwich Cookies   
22          22             0           1  Chocolate Sandwich Cookies   
23          23             0           1  Chocolate Sandwich Cookies   
24          24             0           1  Chocolate Sandwich Cookies   
25          25             0           1  Chocolate Sandwich Cookies   
26          26             0           1  Chocolate Sandwich Cookies   
27          27             0           1  Chocolate Sandwich Cookies   
28          28             0           1  Chocolate Sandwich Cookies   
29          29             0           1  Chocolate Sandwich Cookies   
30          30             0           1  Chocolate Sandwich Cookies   
31          31             0           1  Chocolate Sandwich Cookies   
32          32             0           1  Chocolate Sandwich Cookies   
33          33             0           1  Chocolate Sandwich Cookies   
34          34             0           1  Chocolate Sandwich Cookies   
35          35             0           1  Chocolate Sandwich Cookies   
36          36             0           1  Chocolate Sandwich Cookies   
37          37             0           1  Chocolate Sandwich Cookies   
38          38             0           1  Chocolate Sandwich Cookies   
39          39             0           1  Chocolate Sandwich Cookies   
40          40             0           1  Chocolate Sandwich Cookies   
41          41             0           1  Chocolate Sandwich Cookies   
42          42             0           1  Chocolate Sandwich Cookies   
43          43             0           1  Chocolate Sandwich Cookies   
44          44             0           1  Chocolate Sandwich Cookies   
45          45             0           1  Chocolate Sandwich Cookies   
46          46             0           1  Chocolate Sandwich Cookies   
47          47             0           1  Chocolate Sandwich Cookies   
48          48             0           1  Chocolate Sandwich Cookies   
49          49             0           1  Chocolate Sandwich Cookies   
50          50             0           1  Chocolate Sandwich Cookies   
51          51             0           1  Chocolate Sandwich Cookies   
52          52             0           1  Chocolate Sandwich Cookies   
53          53             0           1  Chocolate Sandwich Cookies   
54          54             0           1  Ch

In [29]:
# Creating flags to assign "freqency" labels to user ID based on median days since prior order
df_ords_prods_merged.loc[df_ords_prods_merged['median_days'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [30]:
df_ords_prods_merged.loc[(df_ords_prods_merged['median_days'] > 10) & (df_ords_prods_merged['median_days'] <= 20),
                     'frequency_flag'] = 'Regular customer'

In [31]:
df_ords_prods_merged.loc[df_ords_prods_merged['median_days'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [32]:
# Check frequencies of spending flags for accuracy
df_ords_prods_merged['frequency_flag'].value_counts(dropna=False)

Frequent customer        22814471
Regular customer          6929754
Non-frequent customer     2689987
Name: frequency_flag, dtype: int64

In [33]:
# Examining the null values
print(df_ords_prods_merged[df_ords_prods_merged.frequency_flag.isnull()])

Empty DataFrame
Columns: [Unnamed: 0, Unnamed: 0_x, product_id, product_name, aisle_id, department_id, prices, Unnamed: 0.2, Unnamed: 0.1, Unnamed: 0_y, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, add_to_cart_order, reordered, _merge, busiest_day, busiest_period_of_day, max_order, loyalty_flag, mean_price, spending_flag, median_days, frequency_flag]
Index: []

[0 rows x 27 columns]


# no null values recorded due to accurate calculations by earlier imputation done

# Exporting the data as pkl for efficiency sake

In [34]:
#marking the outlier values as missing values 
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] >100, 'prices'] = np.nan

In [35]:
#exporting as plk for efficiency sake
df_ords_prods_merged.to_pickle(os.path.join(path,'02. DATA','PREPARED DATA','orders_products_aggregated.pkl'))